// es6-module-packager/dist/lib/locals-compiler.js
"use strict";
var $__TypeError = TypeError,
    $__Object = Object,
    $__getOwnPropertyDescriptor = $__Object.getOwnPropertyDescriptor,
    $__getPrototypeOf = $__Object.getPrototypeOf,
    $__getPropertyDescriptor = function(object, name) {
      while (object !== null) {
        var result = $__getOwnPropertyDescriptor(object, name);
        if (result) return result;
        object = $__getPrototypeOf(object);
      }
      return undefined;
    },
    $__superDescriptor = function(proto, name) {
      if (!proto) throw new $__TypeError('super is null');
      return $__getPropertyDescriptor(proto, name);
    },
    $__superCall = function(self, proto, name, args) {
      var descriptor = $__superDescriptor(proto, name);
      if (descriptor) {
        if ('value'in descriptor) return descriptor.value.apply(self, args);
        if (descriptor.get) return descriptor.get.call(self).apply(self, args);
      }
      throw new $__TypeError("Object has no method '" + name + "'.");
    },
    $__getProtoParent = function(superClass) {
      if (typeof superClass === 'function') {
        var prototype = superClass.prototype;
        if (Object(prototype) === prototype || prototype === null) return superClass.prototype;
      }
      if (superClass === null) return null;
      throw new TypeError();
    },
    $__getOwnPropertyNames = $__Object.getOwnPropertyNames,
    $__getDescriptors = function(object) {
      var descriptors = {},
          name,
          names = $__getOwnPropertyNames(object);
      for (var i = 0; i < names.length; i++) {
        var name = names[i];
        descriptors[name] = $__getOwnPropertyDescriptor(object, name);
      }
      return descriptors;
    },
    $__createClass = function(object, staticObject, protoParent, superClass, hasConstructor) {
      var ctor = object.constructor;
      if (typeof superClass === 'function') ctor.__proto__ = superClass;
      if (!hasConstructor && protoParent === null) ctor = object.constructor = function() {};
      var descriptors = $__getDescriptors(object);
      descriptors.constructor.enumerable = false;
      ctor.prototype = Object.create(protoParent, descriptors);
      Object.defineProperties(ctor, $__getDescriptors(staticObject));
      return ctor;
    },
    $__toObject = function(value) {
      if (value == null) throw $__TypeError();
      return $__Object(value);
    },
    $__spread = function() {
      var rv = [],
          k = 0;
      for (var i = 0; i < arguments.length; i++) {
        var value = $__toObject(arguments[i]);
        for (var j = 0; j < value.length; j++) {
          rv[k++] = value[j];
        }
      }
      return rv;
    };
var GlobalsCompiler = require("es6-module-transpiler").GlobalsCompiler;
var SourceModifier = require("es6-module-transpiler").SourceModifier;
var dirname = require("path").dirname;
var resolve = require("path").resolve;
var LocalsCompiler = function($__super) {
  'use strict';
  var $__proto = $__getProtoParent($__super);
  var $LocalsCompiler = ($__createClass)({
    constructor: function() {
      for (var args = [],
          $__1 = 0; $__1 < arguments.length; $__1++) args[$__1] = arguments[$__1];
      $__superCall(this, $__proto, "constructor", $__spread(args));
      this.options.imports = this.options.imports || {};
    },
    stringify: function() {
      if (!this.moduleName) {
        throw new Error('moduleName must be defiend for locals compiler');
      }
      this.hasDefaultExport = this.exports.reduce(function(acc, item) {
        return acc || item.default;
      }, false);
      var string = this.string.toString();
      this.source = new SourceModifier(string);
      this.map = [];
      var out = this.buildPreamble(this.exports.length > 0);
      this.buildImports();
      this.buildExports();
      out += this.indentLines();
      if (this.exports.length > 0) {
        out += "\n  return __exports__;";
      }
      out += "\n})";
      out += this.buildSuffix();
      out += ";\n";
      return out;
    },
    buildPreamble: function() {
      var out = "",
          dependencyNames = this.dependencyNames;
      if (this.exports.length > 0) {
        out += ("var " + (this.options.imports[this.moduleName] || this.moduleName) + " = ");
      }
      out += "(function(";
      for (var idx = 0; idx < dependencyNames.length; idx++) {
        out += ("__dependency" + (idx + 1) + "__");
        this.map[dependencyNames[idx]] = idx + 1;
        if (!(idx === dependencyNames.length - 1)) out += ", ";
      }
      out += ") {\n";
      out += '  "use strict";\n';
      if (this.exports.length > 0) {
        out += "  var __exports__";
        if (!this.hasDefaultExport) {
          out += " = {}";
        }
        out += ";\n";
      }
      return out;
    },
    buildSuffix: function() {
      var dependencyNames = this.dependencyNames;
      var out = "(";
      for (var idx = 0; idx < dependencyNames.length; idx++) {
        var name = dependencyNames[idx];
        name = this.resolveModuleName(name);
        out += this.options.imports[name] || name;
        if (!(idx === dependencyNames.length - 1)) out += ", ";
      }
      out += ")";
      return out;
    },
    doModuleImport: function(name, dependencyName, idx) {
      return ("var " + name + " = __dependency" + this.map[dependencyName] + "__;\n");
    },
    doDefaultImport: function(name, dependencyName, idx) {
      return ("var " + name + " = __dependency" + this.map[dependencyName] + "__;\n");
    },
    doNamedImport: function(name, dependencyName, alias) {
      return ("var " + alias + " = __dependency" + this.map[dependencyName] + "__." + name + ";\n");
    },
    doDefaultExport: function(identifier) {
      return "__exports__ = ";
    },
    resolveModuleName: function(name) {
      if (/^\.{1,2}\//.test(name)) {
        name = resolve(dirname(this.moduleName), name);
        name = name.replace(/\.js$/, '') + '.js';
      }
      return name;
    }
  }, {}, $__proto, $__super, true);
  return $LocalsCompiler;
}(GlobalsCompiler);
exports["default"] = LocalsCompiler;
